flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/AdaptiveSchedulerTest.java

```diff
@@ -290,6 +303,85 @@ public void testResourceTimeout() throws Exception {
+                        .setRegisterConsumer(
+                                (metric, name, group) -> {
+                                    if (MetricNames.NUM_RESTARTS.equals(name)) {
+                                        numRestartsMetricFuture.complete((Gauge<Integer>) metric);
+                                    }
+                                })
+                        .build();
+
+        final JobGraph jobGraph = createJobGraph();
+
+        final DefaultDeclarativeSlotPool declarativeSlotPool =
+                new DefaultDeclarativeSlotPool(
+                        jobGraph.getJobID(),
+                        new DefaultAllocatedSlotPool(),
+                        ignored -> {},
+                        Time.minutes(10),
+                        Time.minutes(10));
+
+        final Configuration configuration = new Configuration();
+        configuration.set(JobManagerOptions.MIN_PARALLELISM_INCREASE, 1);
+
+        final AdaptiveScheduler scheduler =
+                new AdaptiveSchedulerBuilder(jobGraph, mainThreadExecutor)
+                        .setJobMasterConfiguration(configuration)
+                        .setJobManagerJobMetricGroup(
+                                new JobManagerJobMetricGroup(
+                                        metricRegistry,
+                                        createUnregisteredJobManagerMetricGroup(),
+                                        new JobID(),
+                                        "jobName"))
+                        .setDeclarativeSlotPool(declarativeSlotPool)
+                        .build();
+
+        final Gauge<Integer> numRestartsMetric = numRestartsMetricFuture.get();
+        assertThat(numRestartsMetric.getValue(), is(0));
+
+        scheduler.startScheduling();
+
+        final TaskExecutorGateway taskManagerGateway =
+                new TestingTaskExecutorGatewayBuilder()
+                        .setCancelTaskFunction(
+                                executionAttemptId -> {
+                                    mainThreadExecutor.execute(
+                                            () ->
+                                                    scheduler.updateTaskExecutionState(
+                                                            new TaskExecutionState(
+                                                                    executionAttemptId,
+                                                                    ExecutionState.CANCELED)));
+                                    return CompletableFuture.completedFuture(Acknowledge.get());
+                                })
+                        .createTestingTaskExecutorGateway();
+
+        declarativeSlotPool.offerSlots(
+                createSlotOffersForResourceRequirements(
+                        ResourceCounter.withResource(ResourceProfile.UNKNOWN, 1)),
+                new LocalTaskManagerLocation(),
+                new RpcTaskManagerGateway(taskManagerGateway, JobMasterId.generate()),
```

Comment:
A minor thing: Is there a reason why we use `RpcTaskManagerGateway` instead of `SimpleAckingTaskManagerGateway` here? It wouldn't make a big difference but would reduce the number of instances created by this class...